tools/hidl/AccessMapGenerator.hpp

```diff
@@ -0,0 +1,1003 @@
+        Terms allTerms;
+        for (auto& scopeItem : m_scopes)
+        {
+            if (scopeItem.second.empty())
+                continue;
+
+            Tokens tokens;
+            tokenize(scopeItem.first, scopeItem.second, tokens);
+            Term term;
+            for (std::string& token : tokens)
+            {
+                if (parse(scopeItem.first, token, term))
+                    allTerms.push_front(term);
+            }
+        }
+
+        TermPtrs filteredTerms;
+        bool affirmed = filter(allTerms, filteredTerms);
+        if (!affirmed)
+        {
+            if (m_useDefaultSecurity)
+            {
+                affirmed = filter(m_defaultSecurityTerms, filteredTerms);
+                if (affirmed)
+                    warning("%s/%s is using the default security specification", service, method);
+            }
+            if (!affirmed && m_requireSecurityAffirmation)
+                error("%s/%s is missing an affirmative security specification", service, method);
+        }
+
+        preEntry(filteredTerms.size());
+        for (Term* tp : filteredTerms)
+        {
+            info("Applying %s:%s", tp->feature.c_str(), m_levelMapper.toString(tp->level));
+            entry(tp->feature.c_str(), tp->level);
+        }
+        postEntry();
+
+        return true;
+    }
+
+protected:
+    const ScopeMapper& m_scopeMapper;
+    const LevelMapper& m_levelMapper;
+    const Reporter& m_reporter;
+    Variables m_variables;
+    Scopes m_scopes;
+    bool m_requireSecurityAffirmation = true;
+    bool m_useDefaultSecurity = false;
+    Terms m_defaultSecurityTerms;
+
+private:
+    void normalizeAndSetValue(std::string& str, const char* value, size_t start = 0, size_t end = std::string::npos) const
+    {
+        str.clear();
+        for (size_t idx = start; idx < end && value[idx] != '\0'; idx++)
+        {
+            const char& ch = value[idx];
+            if (!isspace(ch) && ch != '"')
+                str.push_back(ch);
+        }
+    }
+
+    void tokenize(int srcScope, const std::string& fragment, std::list<std::string>& tokens)
+    {
+        debug("tokenizing %s fragment \"%s\"", m_scopeMapper(srcScope), fragment.c_str());
+        static const std::string ignoredCharacters(" \t\n\r\"");
+        std::string token;
+        bool done = false;
+        bool inVariable = false;
+        std::string variable;
+        size_t idx = 0;
+
+        while (!done)
+        {
+            const char& ch = fragment[idx++];
+            switch (ch)
+            {
+            // token delimiters
+            case '\0':
+                done = true;
+                // fall through
+            case ',':
+                if (inVariable)
+                {
+                    warning("%s fragment '%s' contains invalid variable markup (incomplete variable reference '%s')", m_scopeMapper(srcScope), fragment.c_str(), variable.c_str());
+                    token.append("${").append(variable);
```

Comment:
You don't appear to close the { in the token string- is that not required?

Reply:
I intentionally do not assume a close brace. I warn if the close brace is absent but do not assume that one was intended. No combination of $, {, or } was reserved before this change, and this change only reserves the ${…} sequence if … is a defined variable name. Otherwise, the text is accepted as-is.